hal/critical_section_api.h

```diff
@@ -0,0 +1,76 @@
+/** \addtogroup hal */
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_CRITICAL_SECTION_API_H
+#define MBED_CRITICAL_SECTION_API_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * \defgroup hal_critical Critical Section HAL functions
+ * @{
+ */
+
+/**
+  * Mark the start of a critical section
+  *
+  * This function is called directly by core_util_critical_section_enter on
+  * first entrance to a critical section.
+  *
+  * There is a default implementation of this function which will save the
+  * current state of interrupts before disabling them. This implementation can
+  * be found in mbed_critical_section_api.c. This behaviour is can be overridden
+  * on a per platform basis by providing a different implementation within the
+  * correct targets directory.
+  *
+  * The function is only called once per critical section by
+  * core_util_critical_section_enter. When implementing this function for a
+  * platform you must store any state that you intend to alter within this
+  * function so it can be restored when exiting the critical section.
+  *
+  */
+void hal_critical_section_enter(void);
+
+/** Mark the end of a critical section
```

Comment:
Actually could we add some more 'specification' like description? What we expect that function to do (we are using that to make sure that certain operation are not going to be interrupted) and what's going to be tested.